### PR TITLE
HPCC-15965 Reorder inheritance for various significant classes

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
Is this change worth it?

* It reduces the total size of the generated .sos by approximately 1% - some grew slightly, others shrank. 
* Some objects are now smaller - either because fields pack in after the count, or when CInterfaceOf has been used instead.
* Several thunks have been removed from each nextRow() call in thor.  It is unlikely to make a major difference, but it is likely to be beneficial...

@richardkchapman @afishbeck please can you check the changes in your areas - some of them a bit more than reordering the classes.

I will also run the regression suite, but I don't expect it to cause any issues.

**Comment by ghalliday** [comment]
Jira https://track.hpccsystems.com/browse/HPCC-15965 contains more details of the change.

File: common/thorhelper/thortparse.ipp [hunk 1]
```diff
@@ -55,7 +55,7 @@ protected:
 
 //---------------------------------------------------------------------------
 
-class GrammarSymbol : public CInterface, public IMatchedElement
+class GrammarSymbol : public IMatchedElement, public CInterface
```
Comment (richardkchapman) [inline common/thorhelper/thortparse.ipp]: Maybe first public should say implements?

File: dali/base/dasds.cpp [hunk 2]
```diff
@@ -8661,7 +8659,7 @@ void CCovenSDSManager::notifyNode(CServerRemoteTree &node, PDState state)
 
 ///////////////////////
 
-class CDaliSDSServer: public CInterface, public IDaliServer
+class CDaliSDSServer: public IDaliServer, public CInterface
```
Comment (richardkchapman) [inline dali/base/dasds.cpp]: implements ?

**Comment by ghalliday** [comment]
@richardkchapman updated and rebased.

**Pull request merged**